config/config_schema.json

```diff
@@ -341,6 +341,143 @@
+                "clientSecret": {
+                    "description": "Client secret as registered with identity provider",
+                    "type": "string",
+                    "minLength": 1
+                },
+                "scope": {
+                    "description": "Scopes to request from the OpenID Connect server",
+                    "type": "string",
+                    "default": "openid",
+                    "examples": [
+                        "openid",
+                        "openid groups"
+                    ]
+                },
+                "userLookupTable": {
+                    "description": "Path of user lookup table as text file in format <unique user ID> <system user>. If no user lookup is needed, this should be omitted. Example table given in `usertable.txt.stub`",
+                    "type": "string",
+                    "examples": [
+                        "/etc/carta/userlookup.txt"
+                    ]
+                },
+                "groupsField": {
+                    "description": "Name of field containing list of user roles/groups",
+                    "type": "string",
+                    "examples": [
+                        "groups",
+                        "roles"
+                    ]
+                },
+                "requiredGroup": {
+                    "description": "Role to ensure is included among the values in groupsField",
+                    "type": "string",
+                    "examples": [
+                        "carta-users",
+                        "carta-testers"
+                    ]
+                },
+                "localPublicKeyLocation": {
+                    "description": "Path to public key (in PEM format) used for verifying JWTs",
+                    "type": "string",
+                    "examples": [
+                        "/etc/carta/carta_public.pem"
+                    ]
+                },
+                "localPrivateKeyLocation": {
+                    "description": "Path to private key (in PEM format) used for signing JWTs",
+                    "type": "string",
+                    "examples": [
+                        "/etc/carta/carta_private.pem"
+                    ]
+                },
+                "keyAlgorithm": {
+                    "$ref": "#/definitions/keyAlgorithm",
+                    "default": "RS256"
+                },
+                "issuer": {
+                    "description": "Issuer field for JWT",
+                    "type": "string",
+                    "examples": [
+                        "my-carta-server"
+                    ]
+                },
+                "cacheAccessTokenMinValidity": {
+                    "description": "If previously issued an access token from upstream server with at least this many seconds of lifetime remaining, a local token with the previous token's remaining lifetime will be issued instead of performing an upstream query",
```

Comment:
Grammar issues. Maybe this:

"If an access token was previously issued from the upstream server with at least this many seconds of lifetime remaining, a new upstream query will not be performed and a local token with the previous token's remaining lifetime will be issued instead"

Reply:
This BTW is one of the points where I'm wondering if this could pose problems for the scripting token approach / might be worth some additional thoughts from you.  (I'm not sure if you only looked at the schema here or also scanned the code).

Specifically, the approach used here will regenerate the "refresh" token each time a new access token is retrieved from upstream in order to avoid the frontend and controller javascript code separately trying to refresh the access token in quick succession, triggering multiple upstream queries in short succession.

I suspect that ideally we'd like to eliminate the access token retrieval from the frontend codebase entirely.  I'm really not a frontend person (maybe @veggiesaurus can comment?) but I think that it's possible (particular when both frames are from the same origin domain) to have an access token shared from the controller frame to the frontend).  Would imagine that would be a separate set of pull requests to get implemented though

At the moment this caching access token validity of last retrieved access token is basically a way to try to hack around these problems which, if the upstream identity provider will expire a refresh token immediately upon use (as some do), still exist.  This is one problem that to a certain extent still exists here despite the somewhat hackey approach here to try to get around it (i.e. if two requests are running at exact same time): https://auth0.com/docs/secure/tokens/refresh-tokens/refresh-token-rotation#automatic-reuse-detection

(I suspect that a token with `offline_access` scope is likely not to implement rotation + auto-reuse-redetectionion, though I'm not totally sure that that's guaranteed).

Reply:
Also a Q for @veggiesaurus: how important would it be for something like this to roll out right away as part of this (vs separately as originally suggested)?  i.e. do we need a better way of addressing this refresh token rotation / auto-reuse detection?

One thing that makes me hesitant here is that reworking the token handling code to share between controller and frontend would be expected to break the installation if the update of one wasn't paired with the simultaneous update of the other).